lib/vrl/DESIGN.md

```diff
@@ -0,0 +1,219 @@
+dedicated software engineers, because their time is best spent elsewhere.
+
+As with everything, there are exceptions to the rule, and many people in this
+group _are_ highly skilled software engineers, but VRL must capture the largest
+possible segment of this group, and should therefor be limited in complexity.
+
+Therefor, when extending the feature set of VRL, design **the feature for the
+intended target audience**, which will likely mean choosing different trade-offs
+than you'd make if you were to design the feature for your personal needs.
+
+## Language Limits
+
+There are a number of features that we've so far rejected to implement in the
+language. This might change in the future, but there should be a good reason to
+do so.
+
+- modules (see: [#5507][]
+- classes
+- user-defined functions
+- `goto` statements
+- network calls (see [#4517][])
+
+[#5507]: https://github.com/timberio/vector/issues/5507
+[#4517]: https://github.com/timberio/vector/issues/4517#issuecomment-754160338
+
+## Conventions
+
+### Functions
+
+#### Composition
+
+- Favor function composition over single-purpose functions.
+
+- If a problem needs to be solved in multiple steps, consider adding
+  single-purpose functions for each individual step.
+
+- If useability or readability is hurt by composition, favor single-purpose
+  functions.
+
+#### Naming
+
+- Function names are lower-cased (e.g. `round`).
+
+- Multi-name functions use underscore (`_`) as a separator (e.g.
+  `encode_key_value`).
+
+- Favor explicit verbose names over terse ones (e.g.
+  `parse_aws_cloudwatch_log_subscription_message` over `parse_aws_cwl_msg`).
+
+- Use `parse_*` for string decoding functions (e.g. `parse_json` and
+  `parse_grok`).
+
+- Use `encode_*` for string encoding functions (e.g. `encode_base64`).
+
+- Use `to_*` to convert from one type to another (e.g. `to_bool`).
+
+- Use `format_*` for string formatting functions (e.g. `format_timestamp` and
+  `format_number`).
+
+#### Return Types
+
+- Return boolean from `is_*` functions (e.g. `is_string`).
+
+- Return an error when `parse_*` functions fail to decode the string.
+
+#### Mutability
+
+As a general rule, functions never mutate values.
+
+Favor this:
+
+```coffee
+# explicitly assign the parsed JSON to `.message`
+.message = parse_json(.message)
+```
+
+Over this:
+
+```coffee
+# mutate `.message` in place
+parse_json(.message)
+```
+
+There are exceptions to this rule (such as the `del` function), but they are
+limited, and additional exceptions should be well reasoned.
+
+#### Fallibility
+
+- A function must be marked as fallible if it can fail in any way.
+
+- Design a function with the goal of making it infallible.
+
+- But don't hide fallibility for the sake of persuing the previous rule.
+
+- A function implementation can assume it receives the argument type it has
+  defined.
+
+- `parse_*` functions should almost always error when used incorrectly.
+
+- `to_*` functions must never fail.
```

Comment:
This isn't currently true because we decided to have `to_*` functions also take a string, meaning it does _both_ decoding and type conversion.

We can either update this rule to match reality, or (at some point before 1.0) do a breaking change to update the functions to match this rule.

Reply:
💭  would that end up being something like:

```vrl
if is_integer(.field) {
    .field = int(.field)
}
```